### ISSUE TITLE
Launch failed on ARM Mac

### ISSUE THREAD

**Issue by tonyallan** [beeware/briefcase#664]
**Describe the bug**
The application cannot be opened for an unexpected reason

**To Reproduce**
1. Tutorial 1
2. Tutorial 2
3. Tutorial 3

**Expected behavior**
Works as described in the tutorial.

**Screenshots**
```
(beeware-venv) tony@x helloworld % briefcase run

[helloworld] Starting app...

The application cannot be opened for an unexpected reason, error=Error Domain=RBSRequestErrorDomain Code=5 "Launch failed." UserInfo={NSLocalizedFailureReason=Launch failed., NSUnderlyingError=0x60000361eb50 {Error Domain=NSPOSIXErrorDomain Code=153 "Unknown error: 153" UserInfo={NSLocalizedDescription=Launchd job spawn failed}}}

Unable to start app helloworld.
```
**Environment:**
 - Operating System: MacOS Monterey 12.1 (Apple M1 Pro)
 - Python version: 3.8.9
 - Software versions:
   - Briefcase: 0.3.5
   - Toga: 0.3.0.dev20

**Additional context**
Followed tutorials exactly.
Ran code in ~/Downloads folder.

```
python3 -d    
Python 3.8.9 (default, Oct 26 2021, 07:25:53) 
[Clang 13.0.0 (clang-1300.0.29.30)] on darwin
```

**Comment by girivs82** [comment]
Confirm the same issue for me.

**Comment by freakboy3742** [comment]
Thanks for the report. 

Out of interest - if you run a *second* time, does the application start? I've seen an intermittent problem which looks like filesystem metadata taking a couple of seconds to synchronize; if you execute `briefcase run` immediately after it has failed, it works. Unfortunately, I haven't been able to reproduce the problem reliably. though. Does this match your experience?

**Comment by tonyallan** [comment]
Nope

```
(beeware-venv) tony@x helloworld % briefcase run

[helloworld] Starting app...

The application cannot be opened for an unexpected reason, error=Error Domain=RBSRequestErrorDomain Code=5 "Launch failed." UserInfo={NSLocalizedFailureReason=Launch failed., NSUnderlyingError=0x600001341cb0 {Error Domain=NSPOSIXErrorDomain Code=153 "Unknown error: 153" UserInfo={NSLocalizedDescription=Launchd job spawn failed}}}

Unable to start app helloworld.
(beeware-venv) tony@x helloworld % briefcase run

[helloworld] Starting app...

The application cannot be opened for an unexpected reason, error=Error Domain=RBSRequestErrorDomain Code=5 "Launch failed." UserInfo={NSLocalizedFailureReason=Launch failed., NSUnderlyingError=0x600002e7fde0 {Error Domain=NSPOSIXErrorDomain Code=153 "Unknown error: 153" UserInfo={NSLocalizedDescription=Launchd job spawn failed}}}

Unable to start app helloworld.
(beeware-venv) tony@x helloworld % 
```

However `briefcase dev` works with the correct window displayed!

```
[helloworld] Starting in dev mode...
```

**Comment by freakboy3742** [comment]
Thanks for that. I guess that confirms it's a different problem to the one I'm seeing.

The fact that the `dev` command works indicates that the code doesn't have any fundamental problems; so the problem is related to the packaging of the final app binary.

Next debug: Can I get you to manually run the app from the command line - starting from your briefcase folder, run:
```
$ ./macOS/app/Hello\ World/Hello\ World.app/Contents/MacOS/Hello\ World
```
(you may need to tweak the path if you've named the app differently - hopefully the intent is clear. I'm looking to start the binary in the `X.app/Contents/MacOS` folder)

**Comment by tonyallan** [comment]
```
pwd
/Users/tony/Downloads/beeware-tutorial/helloworld

./macOS/app/Hello\ World/Hello\ World.app/Contents/MacOS/Hello\ World
zsh: killed     ./macOS/app/Hello\ World/Hello\ World.app/Contents/MacOS/Hello\ World
```

**Comment by freakboy3742** [comment]
Hrm. Ok - that seems to indicate something fairly fundamental is wrong with the binary.

Next thing to try: MacOS has both an `app` mode, there is also `Xcode` mode. `app` mode uses a pre-compiled bootstrap binary; `Xcode` mode compiles the binary for you. This is usually used if you need to get some custom features into the bootstrap. 

Delete the macOS folder in your project, and run:
```
$ briefcase run macos xcode
```
This will take a little longer to start, because it will be compiling an Xcode project. Does that work any better?

**Comment by tonyallan** [comment]
```
(beeware-venv) tony@lear helloworld % pwd
/Users/tony/Downloads/beeware-tutorial/helloworld

(beeware-venv) tony@lear helloworld % rm -rf macOS 
(beeware-venv) tony@lear helloworld % briefcase run macos xcode

*************************************************************************
** WARNING: Unable to determine the version of Xcode that is installed **
*************************************************************************

   Briefcase will proceed, assuming everything is OK. If you experience
   problems, this is almost certainly the cause of those problems.

   Please report this as a bug at:

     https://github.com/beeware/briefcase/issues/new

   In your report, please including the output from running:

     xcodebuild -version

   from the command prompt.

*************************************************************************


Xcode command line developer tools are not installed.

You should be shown a dialog prompting you to install Xcode and the
command line tools. Select "Install" to install the command line developer
tools.

Re-run Briefcase once that installation is complete.
```

15 minutes later...

```
(beeware-venv) tony@lear helloworld % briefcase run macos xcode

*************************************************************************
** WARNING: Unable to determine the version of Xcode that is installed **
*************************************************************************

   Briefcase will proceed, assuming everything is OK. If you experience
   problems, this is almost certainly the cause of those problems.

   Please report this as a bug at:

     https://github.com/beeware/briefcase/issues/new

   In your report, please including the output from running:

     xcodebuild -version

   from the command prompt.

*************************************************************************


*************************************************************************
** WARNING: Unable to determine the version of Xcode that is installed **
*************************************************************************

   Briefcase will proceed, assuming everything is OK. If you experience
   problems, this is almost certainly the cause of those problems.

   Please report this as a bug at:

     https://github.com/beeware/briefcase/issues/new

   In your report, please including the output from running:

     xcodebuild -version

   from the command prompt.

*************************************************************************


[helloworld] Generating application template...
Using app template: https://github.com/beeware/briefcase-macOS-Xcode-template.git, branch 3.8

[helloworld] Installing support package...
Using support package https://briefcase-support.org/python?platform=macOS&version=3.8
... using most recent revision
Python-3.8-macOS-support.b7.tar.gz already downloaded
Unpacking support package...

[helloworld] Installing dependencies...
Collecting toga-cocoa>=0.3.0.dev20
  Using cached toga_cocoa-0.3.0.dev30-py3-none-any.whl (61 kB)
Collecting toga-core==0.3.0.dev30
  Using cached toga_core-0.3.0.dev30-py3-none-any.whl (489 kB)
Collecting rubicon-objc>=0.3.0
  Using cached rubicon_objc-0.4.2-py3-none-any.whl (58 kB)
Collecting travertino>=0.1.3
  Using cached travertino-0.1.3-py3-none-any.whl (15 kB)
Installing collected packages: travertino, rubicon-objc, toga-core, toga-cocoa
Successfully installed rubicon-objc-0.4.2 toga-cocoa-0.3.0.dev30 toga-core-0.3.0.dev30 travertino-0.1.3

[helloworld] Installing application code...
Installing src/helloworld...

[helloworld] Installing application resources...
Unable to find src/helloworld/resources/helloworld-16.png for 16px application icon; using default
Unable to find src/helloworld/resources/helloworld-32.png for 32px application icon; using default
Unable to find src/helloworld/resources/helloworld-64.png for 64px application icon; using default
Unable to find src/helloworld/resources/helloworld-128.png for 128px application icon; using default
Unable to find src/helloworld/resources/helloworld-256.png for 256px application icon; using default
Unable to find src/helloworld/resources/helloworld-512.png for 512px application icon; using default
Unable to find src/helloworld/resources/helloworld-1024.png for 1024px application icon; using default

[helloworld] Created macOS/Xcode/Hello World

*************************************************************************
** WARNING: Unable to determine the version of Xcode that is installed **
*************************************************************************

   Briefcase will proceed, assuming everything is OK. If you experience
   problems, this is almost certainly the cause of those problems.

   Please report this as a bug at:

     https://github.com/beeware/briefcase/issues/new

   In your report, please including the output from running:

     xcodebuild -version

   from the command prompt.

*************************************************************************


[helloworld] Building XCode project...

objc[86061]: Class AMSupportURLConnectionDelegate is implemented in both /usr/lib/libauthinstall.dylib (0x20d17ab90) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x105b282c8). One of the two will be used. Which one is undefined.
objc[86061]: Class AMSupportURLSession is implemented in both /usr/lib/libauthinstall.dylib (0x20d17abe0) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x105b28318). One of the two will be used. Which one is undefined.
note: Using new build system
note: Planning
note: Build preparation complete
warning: Building targets in manual order is deprecated - check "Parallelize build for command-line builds" in the project editor, or set DISABLE_MANUAL_TARGET_ORDER_BUILD_WARNING in any of the targets in the current build to suppress this warning
Build succeeded.

[helloworld] Built macOS/Xcode/Hello World/build/Release/Hello World.app

[helloworld] Starting app...
```

The app then starts correctly and works ok, however I get an error when I try to stop it...

[bug.txt](https://github.com/beeware/briefcase/files/8056868/bug.txt)

**Comment by freakboy3742** [comment]
Good news on two counts:
1. That seems to verify that the issue is with the stub binary
2. I think I can explain the crash on exit - it looks like a [known macOS M1 issue that was fixed recently](https://github.com/beeware/toga/pull/1406). 

In the short term, you can use the `Xcode` build as a workaround for (1).

If you need an immediate workaround for (2), you can get a local checkout of Toga, and reference that checkout in your pyproject.toml file instead of the formal version markers in the `requires` list.

Apologies for the inconvenience related to this bug. The underlying problem we have is that we don't currently have access to M1 hardware for CI, so we can't really validate at build time that anything will work on M1 hardware. This is something we're trying to resolve.

On an unrelated note - the log you provided had messages about not recognising your Xcode version. Can I ask for the output of `xcodebuild -version` as the message requests?

**Comment by tonyallan** [comment]
I'm just having a play for now.

I'll leave Apple's helpful error messages in the response just in case they are a problem for you...

```
./macOS/Xcode/Hello World/build/Release/Hello World.app
(beeware-venv) tony@lear helloworld % xcodebuild -version
objc[86306]: Class AMSupportURLConnectionDelegate is implemented in both /usr/lib/libauthinstall.dylib (0x20d17ab90) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1084b82c8). One of the two will be used. Which one is undefined.
objc[86306]: Class AMSupportURLSession is implemented in both /usr/lib/libauthinstall.dylib (0x20d17abe0) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1084b8318). One of the two will be used. Which one is undefined.
Xcode 13.2.1
Build version 13C100
```

Let me know if you want me to test anything else.

**Comment by freakboy3742** [comment]
Thanks for those error message - I'll open a separate ticket to get that edge case cleaned up.

**Comment by girivs82** [comment]
My experience is identical to tonyallan's.

**Comment by freakboy3742** [comment]
Good news! I've worked out what is going on here - it's code signing. 

Big Sur on M1 introduced stricter requirements for code signing - *all* apps must now be signed. This isn't enforced on Intel hardware. 

Luckily, there is a relatively painless workaround - once you've run `briefcase build` (or `briefcase run` and then got a failure to start), run:
```
$ briefcase package -p app --adhoc-sign
```

This will apply an adhoc signature to your app, which is enough to get the app running on your own machine. The app won't be distributable until you've signed it with a full signature, but it's enough to get you running locally.

This was also logged as #626 - but I didn't connect that the two problems are the same. The fix will be to incorporate adhoc signing as part of the normal app build process.

**Comment by freakboy3742** [comment]
Gah... this was a misfire. 3187be6 should have closed #668.
